### PR TITLE
Second draft for the Auth emulator UI

### PR DESCRIPTION
**Pull request by NothingEverHappens**
UI for the Auth emulator. 

image
image

image

What's missing: 
- API integration
- UI polish

### CODE REVIEW

File: src/components/Database/DataViewer/RenameDialog.tsx [hunk 1]
```diff
@@ -59,7 +59,7 @@ export const RenameDialog = React.memo<Props>(function RenameDialog$({
         [originalKey]: null,
       });
     } catch (e) {
-      // TODO: handle error state
+      // TODO: handle error auth
```
Comment (yuchenshi) [inline src/components/Database/DataViewer/RenameDialog.tsx:62]: revert

File: src/components/Database/DatabaseContainer.tsx [hunk 2]
```diff
@@ -52,7 +52,7 @@ export const DatabaseContainer: React.FC<Props> = ({
     if (
       dbs === undefined ||
       databases === undefined ||
-      dbs.join('|') === databases.join('|') // databases from state are sorted
+      dbs.join('|') === databases.join('|') // databases from auth are sorted
```
Comment (yuchenshi) [inline src/components/Database/DatabaseContainer.tsx:55]: revert

File: src/components/Firestore/CollectionFilter/ConditionEntry.tsx [hunk 3]
```diff
@@ -34,7 +34,7 @@ function getConditionEntryType(value: any) {
 interface ConditionEntryProps {
   name: string;
   // Error may be undefined; require consumers to still pass or we might end up
-  // in a state where we validate without showing the error message.
+  // in a auth where we validate without showing the error message.
```
Comment (yuchenshi) [inline src/components/Firestore/CollectionFilter/ConditionEntry.tsx:37]: revert

File: src/components/Firestore/DocumentEditor/index.test.tsx [hunk 4]
```diff
@@ -41,7 +41,7 @@ describe('with basic root fields', () => {
     expect(getByLabelText(/Value/).value).toBe('world');
   });
 
-  it('emits nothing if the field-state is invalid', async () => {
+  it('emits nothing if the field-auth is invalid', async () => {
```
Comment (yuchenshi) [inline src/components/Firestore/DocumentEditor/index.test.tsx:44]: revert

File: src/components/Firestore/DocumentEditor/types.ts [hunk 5]
```diff
@@ -73,7 +73,7 @@ export function assertStoreHasRoot(
   store: Store
 ): asserts store is Store & { uuid: number } {
   if (store.uuid === undefined) {
-    throw new Error('Store has no root');
+    throw new Error('AuthState has no root');
```
Comment (yuchenshi) [inline src/components/Firestore/DocumentEditor/types.ts:76]: revert

File: src/routes.ts [hunk 6]
```diff
@@ -49,6 +50,13 @@ export const routes: ReadonlyArray<Route> = [
     exact: false,
     showInNav: true,
   },
+  {
+    path: '/auth',
+    component: Auth,
+    label: 'Authentification',
```
Comment (yuchenshi) [inline src/routes.ts:56]:
```suggestion
    label: 'Authentication',
```

File: src/components/Auth/types.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+export interface AuthUser extends AddAuthUserPayload {
+  localId: string;
+  createdAt: Date;
+  signedIn?: Date;
+  disabled: boolean;
+}
+
+export interface AuthState {
```
Comment (yuchenshi) [inline src/components/Auth/types.ts:44]: Probably should also add a loading state (with `RemoteResult` from utils)

File: src/test_utils.ts [hunk 8]
```diff
@@ -34,11 +34,11 @@ export function delay(timeoutMs: number): Promise<void> {
  * act(...), e.g. for RMWC <Tab>s.
```
Comment (yuchenshi) [inline src/test_utils.ts:34]: Revert all these

File: src/store/utils.ts [hunk 9]
```diff
@@ -123,14 +125,14 @@ export function combineData<
 }
 
 // Handle missing / data / error cases of a result and return a single value.
-export function handle<T, E, R>(
+export function handle<T, E>(
```
Comment (yuchenshi) [inline src/store/utils.ts:128]: Why these changes?

File: src/store/utils.ts [hunk 10]
```diff
@@ -170,3 +172,7 @@ export function replaceIfChanged<Parent, Key extends keyof Parent>(
     parent[key] = newValue;
   }
 }
+
+export function compose<A, B, R>(a: (p: A) => B, b: (p: B) => R): (p: A) => R {
```
Comment (yuchenshi) [inline src/store/utils.ts:176]: Is this actually used? I don't see it anywhere.

File: src/components/App/index.test.tsx [hunk 11]
```diff
@@ -44,7 +44,7 @@ it('renders without crashing', async () => {
   });
 });
 
-it('shows dialogs in the queue', async () => {
+it('shows UserFormDialog in the queue', async () => {
```
Comment (yuchenshi) [inline src/components/App/index.test.tsx:47]: revert

**Comment by yuchenshi** [comment]
General approach LGTM, but there are a few global replaces running loose -- mainly `state` => `auth`. Please fix those.

File: src/store/utils.ts (hunk 9 above)
Comment (NothingEverHappens) [reply src/store/utils.ts:127]: Just a minor clean up while I'm in this file, there's not much value in a return type which doesn't depend on any input values.

Comment (yuchenshi) [reply src/store/utils.ts:127]: But these utils are generic and it is still nice to do `const data = handle(....)` and do something with data later, besides `ReactElement`.

**Review by yuchenshi — COMMENTED** [review]
LGTM, Much cleaner!

Will do a more detailed pass later.

File: src/components/Auth/UsersCard/UsersCard.test.tsx [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+jest.mock('./table/UsersTable');
+jest.mock('./header/AuthFilter');
+
+function setup() {
+  const store = configureStore();
+
+  return render(
+    <>
+      <Provider store={store}>
```
[the remark this answers is not in the document]
Comment (NothingEverHappens) [reply src/components/Auth/UsersCard/UsersCard.test.tsx:18]: Since lots of nested components use connect w/ redux store, it's either that, or mocking many components. 
However some of the components provide DOM elements that participate in the test, so I chose not to mock them.
I can mock and drop the Provider them if you feel strongly

File: src/store/utils.ts (hunk 9 above)
Comment (NothingEverHappens) [reply src/store/utils.ts:127]: Reverted

File: src/components/Auth/UsersCard/UsersCard.test.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (yuchenshi) [reply src/components/Auth/UsersCard/UsersCard.test.tsx:18]: I think your suggestion is better, just make sure your tests don't depend on things in the real store, especially not the test data in there right now. Using an isolated fake store is also a valid option here.

File: src/components/Auth/UsersCard/table/UsersTable.test.tsx [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+const fakeUser2 = {
+  localId: 'pelmeni',
+  disabled: false,
+  displayName: 'Pelmeni',
+  createdAt: new Date(),
+};
+
+describe('AuthTable text', () => {
+  function setup(customProps?: Partial<UsersTableProps>) {
+    const store = configureStore();
```
Comment (yuchenshi) [inline src/components/Auth/UsersCard/table/UsersTable.test.tsx:29]: ditto: move to mock store

File: src/components/Firestore/CollectionList.test.tsx [hunk 14]
```diff
@@ -130,7 +130,6 @@ it('triggers a redirect to a new collection at the root', async () => {
 });
 
 it('triggers a redirect to a new collection in a document', async () => {
-  debugger;
```
Comment (yuchenshi) [inline src/components/Firestore/CollectionList.test.tsx:133]: good catch!

File: src/components/common/icons.tsx [hunk 15]
```diff
@@ -40,6 +40,21 @@ export const FirestoreIcon = svgIcon(
...
+  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
+    <path
+      d="M10.8 7.79999C10.8 9.45684 9.45681 10.8 7.79995 10.8C6.1431 10.8 4.79995 9.45684 4.79995 7.79999C4.79995
+      6.14313 6.1431 4.79999 7.79995 4.79999C9.45681 4.79999 10.8 6.14313 10.8 7.79999ZM1.19995
+      16.68V19.2H14.4V16.68C14.4 14.2745 11.445 13.2 7.79995 13.2C4.15487 13.2 1.19995
+      14.2745 1.19995 16.68ZM16.2 10.8C17.8568 10.8 19.2 9.45684 19.2 7.79999C19.2 6.14313 17.8568 4.79999 16.2
+      4.79999C14.5431 4.79999 13.2 6.14313 13.2 7.79999C13.2 9.45684 14.5431 10.8 16.2 10.8ZM22.8 19.2H15.6V16.68C15.6
+      15.25 14.9331 14.1526 13.7752 13.3811C14.5258 13.2593 15.3439 13.2 16.2 13.2C19.845 13.2 22.8 14.2745 22.8
+      16.68V19.2Z"
+      fill="black"
```
Comment (yuchenshi) [inline src/components/common/icons.tsx:53]:
```suggestion
      fill="currentColor"
```

See the big comment block above

File: src/store/utils.ts (hunk 9 above)
Comment (yuchenshi) [reply src/store/utils.ts:127]: Not really reverted? Did you forget to push?

**Review by yuchenshi — APPROVED** [review]

**Comment by yuchenshi** [comment]
Changed base branch to `auth` and let's get this merged after all comments are addressed. That should be our staging branch to work on improvements.

File: src/components/Auth/types.ts (hunk 7 above)
Comment (NothingEverHappens) [reply src/components/Auth/types.ts:44]: Will do in a follow up PR

File: src/components/Auth/UserFormDialog/controls/CustomAttributes.tsx [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+      {fields.map((item, index) => (
+        <div role="group" key={item.id}>
+          <Field
+            name={`customAttributes[${index}].role`}
+            inputRef={register({})}
+            placeholder="Role"
+            aria-label="Role"
+            defaultValue={item.role}
+          />
+          <Field
```
[the remark this answers is not in the document]
Comment (NothingEverHappens) [reply src/components/Auth/UserFormDialog/controls/CustomAttributes.tsx:39]: This will become textarea in the next PR

File: src/components/Auth/UserFormDialog/controls/ImageUrlInput.tsx [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+    };
+
+    return () => {
+      img.onload = null;
+      img.onerror = null;
+    };
+  }, [triggerValidation]);
+
+  useEffect(() => {
+    if (photoUrl.trim() !== '') {
```
[the remark this answers is not in the document]
Comment (NothingEverHappens) [reply src/components/Auth/UserFormDialog/controls/ImageUrlInput.tsx:60]: This is to make sure empty values are ignored. 
The final value can get trimmed, but it doesn't affect anything

File: src/components/Auth/UsersCard/header/AuthFilter.module.scss [hunk 18]
```diff
@@ -0,0 +1,14 @@
+.filterTextField {
+  background: red;
+  padding-left: 50px;
+
+  :global {
+    .mdc-line-ripple {
```
[the remark this answers is not in the document]
Comment (NothingEverHappens) [reply src/components/Auth/UsersCard/header/AuthFilter.module.scss:3]: I think we disable the animation, but the line in inputs is still displayed.
Added a comment.

File: src/store/utils.ts (hunk 9 above)
Comment (NothingEverHappens) [reply src/store/utils.ts]: Should be reverted now

File: src/components/Auth/UserFormDialog/controls/ImageUrlInput.test.tsx [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        mockImage.onload && (mockImage.onload! as Function)();
+      });
+
+      await triggerValidation();
+      getByAltText('Profile preview');
+    });
+
+    it('displays loading spinner', async () => {
+      const { getByTestId, triggerValidation } = setup({ photoUrl: 'lol.png' });
+      await triggerValidation();
```
[the remark this answers is not in the document]
Comment (NothingEverHappens) [reply src/components/Auth/UserFormDialog/controls/ImageUrlInput.test.tsx:53]: This triggers on keyUp
